aurora/core.py

```diff
@@ -1118,56 +2164,297 @@ def calc_Zeff(self):
         Z = np.arange(Zmax + 1)
         self.delta_Zeff = nz * (Z * (Z - 1))[None, :, None]  # for each charge state
         self.delta_Zeff /= self.ne.T[:, None, :]
+        
+    def calc_taumain(self):
+        (
+            nz,
+            N_mainwall,
+            N_divwall,
+            N_div,
+            N_pump,
+            N_out,
+            N_mainret,
+            N_divret,
+            N_tsu,
+            N_dsu,
+            N_dsul,
+            rcld_rate,
+            rcld_refl_rate,
+            rcld_recl_rate,
+            rcld_impl_rate,
+            rcld_sput_rate,
+            rclb_rate,
+            rcls_rate,
+            rclp_rate,
+            rclw_rate,
+            rclw_refl_rate,
+            rclw_recl_rate,
+            rclw_impl_rate,
+            rclw_sput_rate
+        ) = self.res
+        nz = nz.transpose(2, 1, 0)  # time,nZ,space
+        
+        # factor to account for cylindrical geometry:
+        circ = 2 * np.pi * self.Raxis_cm  # cm
+
+        # calculate total impurity density (summed over charge states)
+        total_impurity_density = np.nansum(nz, axis=1)  # time, space
+
+        # compute total number of particles in the plasma
+        N_main = grids_utils.vol_int(
+            total_impurity_density, self.rvol_grid, self.pro_grid, self.Raxis_cm,
+            rvol_max = None
+        )
+        
+        # compute fluxes leaving the plasma
+        edge_loss = N_tsu * circ
+        limiter_loss = N_dsul * circ
+        parallel_loss = N_dsu * circ
+        flux_out = edge_loss + limiter_loss + parallel_loss
+        
+        # calculate the impurity confinement time in the plasma
+        tau_main = np.divide(N_main, flux_out, out=np.zeros_like(N_main), where=flux_out!=0)
+        
+        return tau_main
```

Comment:
@odstrcilt It is time-dependent, i.e. equal in each time step to the number of particles present in the plasma divided by the flux of particles leaving it. I thought it might be useful to automatically have an automatic calculation of something which resembles a confinement time for the particles in the plasma. i.e. a time scale which would describe the dynamic loss of particles from the plasma if this was also a 0D reservoir, i.e. such flux of particles leaving the plasma = total number of particles in the plasma / tau_main